kube-proxy consider endpoint readiness to delete UDP stale conntrack entries

aojea opened the pull request:
/kind bug
/kind regression


#### What this PR does / why we need it:

1. Create an UDP Service
2. Client Pod sending traffic to the UDP service
3. Create an UDP server associated to the Service created in 1. with an init container that sleeps for some time

The init container makes that the server pod is not ready, however, the endpoint slices are created, it is just
that the Endpoint conditions Ready is false.
If the kube-proxy conntrack logic doesn't check readiness, it will delete the conntrack entries for the UDP server
when the endpoint slice has been created, however, the iptables rules will not installed until at least one
endpoint is ready. If some traffic arrives to since kube-proxy clear the entries (see the endpoint slice) and
installs the corresponding iptables rules (the endpoint is ready), a conntrack entry will be generated blackholing
subsequent traffic.


Fixes #105657

```release-note
fix a 1.22 kube-proxy regression on UDP services because the logic to detect stale connections was not considering if the endpoint was ready.
```

Commit:
kube-proxy: fix stale detection logic

The logic to detect stale endpoints was not assuming the endpoint
readiness.

We can have stale entries on UDP services for 2 reasons:
- an endpoint was receiving traffic and is removed or replaced
- a service was receiving traffic but not forwarding it, and starts
to forward it.

Add an e2e test to cover the regression

aojea commented:
/sig network

aojea commented:
ok, the test reproduces the issue

> Kubernetes e2e suite: [sig-network] Conntrack should be able to preserve UDP traffic when initial unready endpoints get ready expand_less	1m40s
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/conntrack.go:293
Nov  5 00:08:20.758: Failed to connect to backend pod
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113

aojea commented:
/assign @thockin @danwinship

danwinship commented on pkg/proxy/endpoints.go:
```diff
@@ -561,8 +567,21 @@ func detectStaleConnections(oldEndpointsMap, newEndpointsMap EndpointsMap, stale
+		epReady := 0
+		for _, ep := range epList {
+			if ep.IsReady() {
+				epReady++
```
So... this doesn't take `ProxyTerminatingEndpoints` into account...

Consider the case where an `externalTrafficPolicy: Local` service has a single Serving-Terminating endpoint. Connections come in to that endpoint's node and are accepted and processed by the terminating pod. Then a new endpoint starts up and becomes Ready. Given the code here, that would be interpreted as "the service went from 0 endpoints to non-0 endpoints", and so the node with the Serving-Terminating endpoint would flush all conntrack entries for the service, breaking the existing connections to the Serving-Terminating pod.

(Also, this patch changes the rules for `staleServices`, but there are terminating endpoints problems with `staleEndpoints` too; we used to delete conntrack entries to endpoints as soon as the endpoint become non-ready, but now we don't delete them until the pod is fully deleted...)

aojea replied:
> So... this doesn't take `ProxyTerminatingEndpoints` into account...

this is a regression that needs to be backported  and ProxyTerminatingEndpoints is an alpha feature. Also, after the analysis you did in your related PR, I don't think that is easy to solve both problems easily at the same time :sweat_smile: 

>  breaking the existing connections to the Serving-Terminating pod.

it is UDP, in the sense that is not breaking the connection per se, the new packet will create a new entry looking at the iptables rules, that should still exist ... is less performant because but not a big deal

> (Also, this patch changes the rules for staleServices, but there are terminating endpoints problems with staleEndpoints too; we used to delete conntrack entries to endpoints as soon as the endpoint become non-ready, but now we don't delete them until the pod is fully deleted...)

that is fixed by the `Equal` change to take into consideration Ready https://github.com/kubernetes/kubernetes/pull/106163/files#r743336980

danwinship replied:
> it is UDP, in the sense that is not breaking the connection per se, since it is connectionless

UDP is connectionless at L4, but not necessarily at L7. That's the main reason UDP conntrack records exist. Eg, anything using Datagram TLS (like QUIC / HTTP/3) won't survive being switched to a different endpoint mid-communication, because the new endpoint won't have the encryption key it needs.

aojea replied:
:eyes: I can't argue against that, but seems we are going to have some fun soon :smile:

aojea commented on pkg/proxy/endpoints.go:
```diff
@@ -121,7 +121,13 @@ func (info *BaseEndpointInfo) Port() (int, error) {
 
 // Equal is part of proxy.Endpoint interface.
 func (info *BaseEndpointInfo) Equal(other Endpoint) bool {
-	return info.String() == other.String() && info.GetIsLocal() == other.GetIsLocal()
+	return info.String() == other.String() &&
```
thanks

aojea replied:
adding only the ready condition to the Equal comparison to keep this targeted to the problem at hand

danwinship commented on pkg/proxy/endpoints.go:
```diff
@@ -536,13 +538,22 @@ func (em EndpointsMap) getLocalReadyEndpointIPs() map[types.NamespacedName]sets.
+			}
 			stale := true
+			// Check if the endpoint has changed, including if it went from ready to not ready.
+			// Ff it did change stale entries for the old endpoint has to be cleared.
```
```suggestion
			// If it did change stale entries for the old endpoint has to be cleared.
```

danwinship commented on pkg/proxy/endpoints.go:
```diff
@@ -536,13 +538,22 @@ func (em EndpointsMap) getLocalReadyEndpointIPs() map[types.NamespacedName]sets.
 		}
 
 		for _, ep := range epList {
+			// if the old endpoints wasn't ready is not possible to have stale entries
```
```suggestion
			// if the old endpoint wasn't ready is not possible to have stale entries
```

danwinship reviewed (commented):
/approve
just some typos. (I would have let the "endpoints"/"endpoint" slide, but the "Ff" would annoy me :slightly_smiling_face:)

aojea commented:
> Kubernetes e2e suite: [sig-node] PreStop graceful pod terminated should wait until preStop hook completes the process expand_more	

:eyes: 


/test pull-kubernetes-e2e-gce-ubuntu-containerd
--


<br class="Apple-interchange-newline">

aojea commented:
ok 2/2 ?
```


Kubernetes e2e suite: [sig-node] PreStop graceful pod terminated should wait until preStop hook completes the process expand_less
```

aojea commented:
> Nov  5 15:59:37.306: INFO: At 2021-11-05 15:58:51 +0000 UTC - event for pod-prestop-hook-06a9dd06-d340-40c3-8676-a84f44b9a683: {kubelet e2e-b6818f6f36-a7d53-minion-group-jrkd} FailedPreStopHook: Exec lifecycle hook ([sh -c while true; do echo preStop; sleep 1; done]) for Container "nginx" in Pod "pod-prestop-hook-06a9dd06-d340-40c3-8676-a84f44b9a683_prestop-9129(22f8a03f-f8bf-49e2-814e-548404254983)" failed - error: command 'sh -c while true; do echo preStop; sleep 1; done' exited with 137: , message: "preStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\npreStop\n"

137 = OOM?

the same test passed in the other jobs, I can't find the relation with this PR
/test pull-kubernetes-e2e-gce-ubuntu-containerd

I will run it some times more

aojea commented:
a different one this time?
```
Kubernetes e2e suite: [sig-network] Proxy version v1 should proxy logs on node with explicit kubelet port using proxy subresource  expand_more
``` 

can it be related to fc85bb29d2fbb844a5a4e37cbc851431deb608db ?

aojea commented:
> a different one this time?
> 
> ```
> Kubernetes e2e suite: [sig-network] Proxy version v1 should proxy logs on node with explicit kubelet port using proxy subresource  expand_more
> ```
> 
> can it be related to [fc85bb2](https://github.com/kubernetes/kubernetes/commit/fc85bb29d2fbb844a5a4e37cbc851431deb608db) ?

so, it was that commit, @danwinship , this is ready to merge and backport, I squashed the e2e test with the kube-proxy changes and removed the commit that was causing issues with the e2e framework

neolit123 commented:
/remove-area kubeadm
/remove-sig cluster-lifecycle

danwinship commented:
/lgtm
/approve

fedebongio commented:
/triage accepted
